common/json.c

```diff
@@ -301,6 +301,90 @@ const jsmntok_t *json_get_arr(const jsmntok_t tok[], size_t index)
 	return NULL;
 }
 
+/* Finds the end of a JSMN token.  */
+static const jsmntok_t *
+find_jsmn_bounds(const jsmntok_t *toks, const jsmntok_t *toks_end, int size)
+{
+	const jsmntok_t *p = toks;
+	const jsmntok_t *e = toks_end;
+	int i;
+
+	for (i = 0; i < size && p < e; ++i) {
+		if (p->size == 0)
+			++p;
+		else
+			p = find_jsmn_bounds(p + 1, e, p->size);
+	}
+
+	return p;
+}
+
+/* Validate that the JSMN result is not weird, e.g.
+ { "" "" "" }, which JSMN will accept incorectly.
+*/
+static bool
+validate_jsmn_result(const jsmntok_t *toks, const jsmntok_t *toks_end)
```

Comment:
This is supposedly not only true for results but arbitrary JSON-encoded objects.

Reply:
Well it protects against certain infelicities of the JSMN parser.  There are invalid JSON that JSMN accepts; this function is intended to filter *some* of those out (for example, it will still accept objects with duplicate keys).  Presumably a more feature-complete JSON parser would outright reject those invalid JSON.

Reply:
No, I meant this check should be applied to all JSON objects we handle, whether they are requests or results. The name of the function suggests it's for results only.

Reply:
The other objects we handle tend to be in `json_stream` or `json_out` forms, which have their own issues, but are *not* issues *this* function can detect.  This function handles LIBJSMN infelicities, specifically the `jsmn_parse` weirdnesses; other issues in `json_stream` / `json_out` are not handled by this function.  So it makes no sense to apply this to other JSON objects; it should only apply to the result of `jsmn_parse`, because that is *all* it actually checks.

In theory, if we had a decent library instead of a small fast one we would not need this function.  It is strictly due to the use of LIBJSMN, and it is the output result of LIBJSMN that it checks.

How about `validate_jsmn_parse_output` instead?

Reply:
That'd be perfect, and you're right, we mostly use jsmn for results.